riscv/src/register/macros.rs

```diff
@@ -600,3 +600,510 @@ macro_rules! clear_pmp {
+    ) => {
+         $crate::read_only_csr_field!(
+             $ty,
+             $(#[$field_doc])+
+             $field,
+             $(#[$try_field_doc])+
+             $try_field,
+             range: $bit_start ..= $bit_end,
+         );
+
+         $crate::write_only_csr_field!(
+             $ty,
+             $(#[$set_field_doc])+
+             $set_field,
+             $(#[$try_set_field_doc])+
+             $try_set_field,
+             range: $bit_start ..= $bit_end,
+         );
+    };
+
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident,
+     $(#[$set_field_doc:meta])+
+     $set_field:ident,
+     range: [$bit_start:literal : $bit_end:literal]$(,)?
+    ) => {
+        $crate::read_only_csr_field!(
+            $ty,
+            $(#[$field_doc])+
+            $field: [$bit_start : $bit_end],
+        );
+
+        $crate::write_only_csr_field!(
+            $ty,
+            $(#[$set_field_doc])+
+            $set_field: [$bit_start : $bit_end],
+        );
+    };
+
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident,
+     $(#[$try_field_doc:meta])+
+     $try_field:ident,
+     $(#[$set_field_doc:meta])+
+     $set_field:ident,
+     $(#[$field_ty_doc:meta])+
+     $field_ty:ident {
+         range: [$field_start:literal : $field_end:literal],
+         default: $default_variant:ident,
+         $($variant:ident = $value:expr$(,)?)+
+     }$(,)?
+    ) => {
+        $crate::csr_field_enum!(
+            $(#[$field_ty_doc])+
+            $field_ty {
+                range: [$field_start : $field_end],
+                default: $default_variant,
+                $($variant = $value,)+
+            },
+         );
+
+         $crate::read_only_csr_field!(
+             $ty,
+             $(#[$field_doc])+
+             $field,
+             $(#[$try_field_doc])+
+             $try_field,
+             $field_ty,
+             range: [$field_start : $field_end],
+         );
+
+         $crate::write_only_csr_field!(
+             $ty,
+             $(#[$set_field_doc])+
+             $set_field,
+             $field_ty,
+             range: [$field_start : $field_end],
+         );
+    };
+}
+
+/// Defines field accesor functions for a read-only CSR type.
+#[macro_export]
+macro_rules! read_only_csr_field {
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident: $bit:literal$(,)?) => {
+        const _: () = assert!($bit < usize::BITS);
+
+        impl $ty {
+            $(#[$field_doc])+
+            #[inline]
+            pub fn $field(&self) -> bool {
+                $crate::bits::bf_extract(self.bits, $bit, 1) != 0
+            }
+        }
+    };
+
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident,
+     $(#[$try_field_doc:meta])+
+     $try_field:ident,
+     range: $bit_start:literal..=$bit_end:literal$(,)?) => {
+        const _: () = assert!($bit_end < usize::BITS);
+        const _: () = assert!($bit_start <= $bit_end);
```

Comment:
Why not only `<`? to ensure that there are at least two bits

```suggestion
        const _: () = assert!($bit_start < $bit_end);
```

Reply:
I included the `<=` because it is technically valid, even if it probably would only ever be a typo, e.g. user's would probably want the single-bit variant if they were defining a field where `start == end`.

I can change it to `<`.

Reply:
I included the `<=` because it is technically valid, even if it probably would only ever be a typo, e.g. users would probably want the single-bit variant if they were defining a field where `start == end`.

I can change it to `<`.